### PR TITLE
Fx with meta

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR simplies and improves the way tracing works with torch.fx.

Instead of recording concrete values via a forward pass on the original model, metadata is attached to the proxies, either tensors on the `meta` device (which saves us from making actual computations, only shape inference is performed) or any other type such as `torch.Size` and builtin types. On top of allowing to trace very big models, this gives much more flexibility and should allow to support many new architectures.

A big thanks to @jamesr66a as he was the one who provided the basis for tracing with meta tensors, and I simply extended what was already done to our purposes.

@jamesr66a @pbelevich I would love your review and feedbacks!

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
Not sure why the changes in type annotations of Electra, GPT-Neo and RoBERTa are included in this PR, they should go in a separate PR that's not focused on ONNX.

The actual code changes are way too complex for me to fully parse, so I'll trust you have tested them on all existing models. Make sure you run all slow ONNX tests before merging this PR though. I'd like to know why one of the tests is removed however.

File: src/transformers/models/electra/modeling_electra.py
Comment (michaelbenayoun) [reply src/transformers/models/electra/modeling_electra.py:853]: Yes it is. Basically the missing annotation makes tracing fail for some reason.
I figured that it would be nice to have it so I added those.

File: src/transformers/utils/fx.py
Comment (michaelbenayoun) [reply src/transformers/utils/fx.py:320]: I am not sure I will need this later, that's why I kept it like that.

File: tests/test_modeling_common.py
Comment (michaelbenayoun) [reply tests/test_modeling_common.py:702]: My bad, thanks for catching this!

File: src/transformers/utils/fx.py
Comment (michaelbenayoun) [reply src/transformers/utils/fx.py:146]: I don't think that it's an issue here, because all the asserts are under a `try` block, if one of them raises an exception, it is handled later in the function and a [proper message is being raised](https://github.com/michaelbenayoun/transformers/blob/fx_with_meta/src/transformers/utils/fx.py#L589). Wdty?

File: src/transformers/models/electra/modeling_electra.py
Comment (michaelbenayoun) [reply src/transformers/models/electra/modeling_electra.py:853]: Also, it is for torch.fx, not ONNX.

File: src/transformers/utils/fx.py
Comment (sgugger) [reply src/transformers/utils/fx.py]: Asserts can be deactivated when executing code, so your `try` block will always pass. That's why we use Python errors everywhere else in the codebase.

Comment (michaelbenayoun) [reply src/transformers/utils/fx.py]: Alright, gotcha, I replaced the assertions by explicit exceptions.

**Comment by michaelbenayoun** [comment]
@sgugger is it ready now?

**Review by sgugger — APPROVED** [review]
On my side, yes, but let's wait for @LysandreJik approval as well as this is a lot of change.

**Review by LysandreJik — APPROVED** [review]
Big change! Superficially reviewed, but it looks good to me. I haven't tried it out myself.

Should additional tests be implemented to ensure that everything works as intended? If you think current tests provide enough coverage, happy to merge it like this.

File: src/transformers/utils/fx.py
Comment (jamesr66a) [inline src/transformers/utils/fx.py:343]: This line causes the following code to be dead. Is this intended?

File: src/transformers/utils/fx.py
Comment (jamesr66a) [inline src/transformers/utils/fx.py:339]: It seems like we cannot get concrete shape information with this implementation. I've tried, for example, `GPT2ModelTest::test_torch_fx` and it does not trace properly. Implementing shape and size as the following works:

```
    @property
    def shape(self):
        if hasattr(self, '_tensor_meta') and self._tensor_meta is not None:
            return self._tensor_meta.shape
        return self.tracer.create_proxy('call_function', builtins.getattr, (self, 'shape'), {})

    def size(self, dim=None):
        if hasattr(self, '_tensor_meta') and self._tensor_meta is not None:
            return self._tensor_meta.size(*[dim] if dim else [])
        return self.tracer.create_proxy('call_method', 'size', (self, dim) if dim else (self,), {})
```

Comment (jamesr66a) [inline src/transformers/utils/fx.py:339]: It seems like we cannot get concrete shape information with this implementation. I've tried, for example, `GPT2ModelTest::test_torch_fx` and it does not trace properly. Implementing shape and size as the following works:

```
    @property
    def shape(self):
        if hasattr(self, '_tensor_meta') and self._tensor_meta is not None:
            return self._tensor_meta.shape
        return self.tracer.create_proxy('call_function', builtins.getattr, (self, 'shape'), {})

    def size(self, dim=None):
        if hasattr(self, '_tensor_meta') and self._tensor_meta is not None:
            return self._tensor_meta.size(*[dim] if dim else [])
        return self.tracer.create_proxy('call_method', 'size', (self, dim) if dim else (self,), {})
```

Comment (jamesr66a) [inline src/transformers/utils/fx.py:339]: This implementation does not provide any way to get concrete shape information from the MetaTensor, and thus almost all of the tests are failing because of errors passing tensor shapes to indexing (when `is_torch_fx_available = True`)

Comment (jamesr66a) [reply src/transformers/utils/fx.py:339]: Failing tests for reference:

```
FAILED tests/models/albert/test_modeling_albert.py::AlbertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/albert/test_modeling_albert.py::AlbertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/albert/test_modeling_albert.py::AlbertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/albert/test_modeling_albert.py::AlbertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/bert/test_modeling_bert.py::BertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/bert/test_modeling_bert.py::BertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/bert/test_modeling_bert.py::BertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/bert/test_modeling_bert.py::BertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/distilbert/test_modeling_distilbert.py::DistilBertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/distilbert/test_modeling_distilbert.py::DistilBertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __...
FAILED tests/models/distilbert/test_modeling_distilbert.py::DistilBertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/distilbert/test_modeling_distilbert.py::DistilBertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __...
FAILED tests/models/electra/test_modeling_electra.py::ElectraModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/electra/test_modeling_electra.py::ElectraModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ m...
FAILED tests/models/electra/test_modeling_electra.py::ElectraModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/electra/test_modeling_electra.py::ElectraModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ m...
FAILED tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/gpt_neo/test_modeling_gpt_neo.py::GPTNeoModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/gpt_neo/test_modeling_gpt_neo.py::GPTNeoModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ me...
FAILED tests/models/gptj/test_modeling_gptj.py::GPTJModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/gptj/test_modeling_gptj.py::GPTJModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/megatron_bert/test_modeling_megatron_bert.py::MegatronBertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __inde...
FAILED tests/models/megatron_bert/test_modeling_megatron_bert.py::MegatronBertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or ha...
FAILED tests/models/megatron_bert/test_modeling_megatron_bert.py::MegatronBertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __inde...
FAILED tests/models/megatron_bert/test_modeling_megatron_bert.py::MegatronBertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or ha...
FAILED tests/models/mobilebert/test_modeling_mobilebert.py::MobileBertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/mobilebert/test_modeling_mobilebert.py::MobileBertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __...
FAILED tests/models/mobilebert/test_modeling_mobilebert.py::MobileBertModelTest::test_torch_fx - TypeError: slice indices must be integers or None or have an __index__ method
FAILED tests/models/mobilebert/test_modeling_mobilebert.py::MobileBertModelTest::test_torch_fx_output_loss - TypeError: slice indices must be integers or None or have an __...
FAILED tests/models/roberta/test_modeling_roberta.py::RobertaModelTest::test_torch_fx - AssertionError: Couldn't trace module.
FAILED tests/models/roberta/test_modeling_roberta.py::RobertaModelTest::test_torch_fx_output_loss - AssertionError: Couldn't trace module.
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_torch_fx - AssertionError: Couldn't trace module.
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_torch_fx_output_loss - AssertionError: Couldn't trace module.
```

**Pull request merged**
